azure/functions/decorators/function_app.py

```diff
@@ -1326,140 +1463,104 @@ def decorator():
-        fields to include in the binding json.
+        :return: List of functions in the function app.
+        """
+        return [function_builder.build(self.auth_level) for function_builder
+                in self._function_builders]
 
-        :return: Decorator function.
+    def register_functions(self, function_container: DecoratorApi) -> None:
+        """Register a list of functions in the function app.
+
+        :param function_container: Instance extending :class:`DecoratorApi`
+        which contains a list of functions.
         """
+        if isinstance(function_container, FunctionRegister):
+            raise TypeError('functions can not be type of FunctionRegister!')
+        self._function_builders.extend(function_container._function_builders)
 
-        @self._configure_function_builder
-        def wrap(fb):
-            def decorator():
-                nonlocal kwargs
-                if type == HTTP_TRIGGER:
-                    if kwargs.get('auth_level', None) is None:
-                        kwargs['auth_level'] = self.auth_level
-                    if 'route' not in kwargs:
-                        kwargs['route'] = None
-                fb.add_trigger(
-                    trigger=GenericTrigger(
-                        name=arg_name,
-                        type=type,
-                        data_type=parse_singular_param_to_enum(data_type,
-                                                               DataType),
-                        **kwargs))
-                return fb
 
-            return decorator()
+class FunctionApp(FunctionRegister, TriggerApi, BindingApi):
+    """FunctionApp object used by worker function indexing model captures
+    user defined functions and metadata.
 
-        return wrap
+    Ref: https://aka.ms/azure-function-ref
+    """
 
-    def event_grid_trigger(self,
-                           arg_name: str,
-                           data_type: Optional[
-                               Union[DataType, str]] = None,
-                           **kwargs) -> Callable:
+    def __init__(self,
+                 http_auth_level: Union[AuthLevel, str] = AuthLevel.FUNCTION):
+        """Constructor of :class:`FunctionApp` object.
+
+        :param http_auth_level: Determines what keys, if any, need to be
+        present
+        on the request in order to invoke the function.
         """
-        The event_grid_trigger decorator adds
-        :class:`EventGridTrigger`
-        to the :class:`FunctionBuilder` object
-        for building :class:`Function` object used in worker function
-        indexing model. This is equivalent to defining event grid trigger
-        in the function.json which enables function to be triggered to
-        respond to an event sent to an event grid topic.
-        All optional fields will be given default value by function host when
-        they are parsed by function host.
+        super().__init__(auth_level=http_auth_level)
```

Comment:
we have to think for the future - if auth_level is set and no HTTP function is registered - we should give some sort of a warning - just in case. Not for this PR.

Reply:
will add a task for this